### PR TITLE
Fix drivers convert blob2pnguri

### PR DESCRIPTION
**Pull request by bhavsarpratik**

### CODE REVIEW

File: jina/drivers/convert.py [hunk 1]
```diff
@@ -8,6 +8,7 @@
 import urllib.request
 import zlib
 
+from PIL import Image
```
Comment (JoanFM) [inline jina/drivers/convert.py:11]: Better move this import inside the function needing it

File: jina/drivers/convert.py [hunk 2]
```diff
@@ -136,8 +138,27 @@ def png_pack(png_tag, data):
...
-        arr.uri = self.png_convertor(arr)
+    @staticmethod
+    def image_to_byte_array(image:Image, format:str):
+        import io
+        img_byte_arr = io.BytesIO()
+        image.save(img_byte_arr, format=format)
+        img_byte_arr = img_byte_arr.getvalue()
+        return img_byte_arr
+
+    def png_convertor_nd(self, arr: np.array):
```
Comment (JoanFM) [inline jina/drivers/convert.py:149]: It is weird that we are saying that it is nd but we are converting and interpreting as an RGB image

Comment (JoanFM) [reply jina/drivers/convert.py:149]: Better to say 3d and that is it

**Comment by hanxiao** [comment]
unit test is required

**Review by JoanFM — CHANGES_REQUESTED** [review]

File: jina/drivers/convert.py (hunk 2 above)
Comment (bhavsarpratik) [reply jina/drivers/convert.py:149]: I think it handles greyscale 2d but need to confirm it. Will change it to 3d for now.

Comment (JoanFM) [reply jina/drivers/convert.py:149]: well then at least add exception if the shape is not properly handled and tested

File: jina/drivers/convert.py (hunk 1 above)
Comment (bhavsarpratik) [reply jina/drivers/convert.py:11]: I have to keep it outside for argument type info for image_to_byte_array. Should I remove type hint from the argument of image_to_byte_array? 

I have kept PIL inside the function in png_convertor3d

Comment (JoanFM) [reply jina/drivers/convert.py:11]: if using for type hint you can do:

```python
if False:
   from PIL import Image
```

**Comment by bhavsarpratik** [comment]
I think our test has failed due to the `timeout-minutes: 30` in actions yaml. Should we increase it? To what value? Should I push this change as a separate PR or in this PR?

**Comment by JoanFM** [comment]
> I think our test has failed due to the `timeout-minutes: 30` in actions yaml. Should we increase it? To what value? Should I push this change as a separate PR or in this PR?

I think sometimes fails but it tends to be good. But we do not have a new test for the added cases right?

**Comment by bhavsarpratik** [comment]
No new test is added for now

File: jina/drivers/convert.py [hunk 3]
```diff
@@ -146,19 +146,28 @@ def image_to_byte_array(image:Image, format:str):
         img_byte_arr = img_byte_arr.getvalue()
         return img_byte_arr
 
-    def png_convertor_nd(self, arr: np.array):
+    def png_convertor(self, arr: np.array):
         from PIL import Image
         arr = arr.astype(np.uint8)
-        im = Image.fromarray(arr).convert('RGB')
+        if arr.shape == 1:
```
Comment (JoanFM) [inline jina/drivers/convert.py:152]: len(arr.shape)?

Comment (bhavsarpratik) [reply jina/drivers/convert.py:152]: Oopsie

File: jina/drivers/convert.py [hunk 4]
```diff
@@ -139,35 +140,36 @@ def png_pack(png_tag, data):
...
-    def image_to_byte_array(image:Image, format:str):
+    def image_to_byte_array(image: 'Image', format: str):
         import io
         img_byte_arr = io.BytesIO()
         image.save(img_byte_arr, format=format)
         img_byte_arr = img_byte_arr.getvalue()
         return img_byte_arr
 
     def png_convertor(self, arr: np.array):
+        import numpy as np
```
Comment (JoanFM) [inline jina/drivers/convert.py:151]: Numpy is one of our key dependency and it can be put on top

Comment (bhavsarpratik) [reply jina/drivers/convert.py:151]: It is also kept at top. I have kept it again inside for performance as this function can be called many times. Should I remove it from the function?

Comment (JoanFM) [reply jina/drivers/convert.py:151]: why is it better for performance to keep it inside the function?

Comment (bhavsarpratik) [reply jina/drivers/convert.py:151]: Han told about this earlier. 
https://github.com/jina-ai/jina/issues/373#issuecomment-623483847

Comment (JoanFM) [reply jina/drivers/convert.py:151]: But this is about running simple `jina` commands (for instance doing `jina --help` and this kind of things) to run fast by not doing extra imports, but for the function to run fast it does not affect (or it could only be harmful). Plus, if it is alrady imported on the top.

Comment (bhavsarpratik) [reply jina/drivers/convert.py:151]: Okay, would like to understand why are we importing inside the functions over here?
https://github.com/jina-ai/jina-hub/blob/master/encoders/nlp/TransformerTorchEncoder/__init__.py

To confirm if importing inside helps, earlier I had tried a function with numpy import inside and outside. I think it was slightly faster when I kept it inside. I am also new to this concept of importing inside. 

As per this, it's cached the first time the function is used. So it doesn't make it slower by importing every time.
https://stackoverflow.com/questions/3095071/in-python-what-happens-when-you-import-inside-of-a-function

Comment (JoanFM) [reply jina/drivers/convert.py:151]: It should not affect, we import inside when it is not a basic dependency (so that the fact that only importing that file do not crash jina), `numpy` is a basic dependencya and therefore it is okey

Comment (JoanFM) [reply jina/drivers/convert.py:151]: but the main point here, is that u are importing both inside and outside. And since it is numpy let's keep it outside

Comment (bhavsarpratik) [reply jina/drivers/convert.py:151]: Yes I agree with the practice.

File: tests/unit/drivers/test_convert_driver.py [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+
+
+@pytest.mark.parametrize('data', test_data)
+def test_blob2pnguri_driver(data):
+    doc, mode = data
+    width, height = 28, 32
+
+    driver = Blob2PngURI(target='uri', width=width, height=height)
+    driver._apply(doc)
+    Image.frombytes(mode, (width, height), doc.uri.encode()) # just to check if the data is enough for the image recreation
```
Comment (JoanFM) [inline tests/unit/drivers/test_convert_driver.py:28]: some assertion at least about the resulting size?

**Review by JoanFM — CHANGES_REQUESTED** [review]

File: tests/unit/drivers/test_convert_driver.py (hunk 5 above)
Comment (bhavsarpratik) [reply tests/unit/drivers/test_convert_driver.py:28]: I had actually kept the assertion but felt it useless since the shape is given for recreating the image. If it is not able to create it from the byte data, it throws the error. Also, I observed that this is not very reliable for exact checking. I can have input image of some shape and then have this recreated of some other shape and it still works. So this recreation is just to check if the uri data exists and is enough.

File: jina/drivers/convert.py [hunk 6]
```diff
@@ -136,7 +139,35 @@ def png_pack(png_tag, data):
...
+        img_byte_arr = img_byte_arr.getvalue()
+        return img_byte_arr
+
+    def png_convertor(self, arr: np.array):
+        from PIL import Image
+
+        arr = arr.astype(np.uint8)
+
+        if len(arr.shape) == 1:
+            # im = Image.fromarray(arr.reshape((self.height, self.width))).convert('L')
```
Comment (JoanFM) [inline jina/drivers/convert.py:156]: remove comment

File: tests/unit/drivers/test_convert_driver.py (hunk 5 above)
Comment (JoanFM) [reply tests/unit/drivers/test_convert_driver.py:28]: and testing that URI2Buffer returns the same value as in Blob would make sense?

File: jina/drivers/convert.py (hunk 6 above)
Comment (bhavsarpratik) [reply jina/drivers/convert.py:156]: This is our replacement code for the png_converter_1d. Will remove and save it somewhere

File: tests/unit/drivers/test_convert_driver.py (hunk 5 above)
Comment (bhavsarpratik) [reply tests/unit/drivers/test_convert_driver.py:28]: The blob is converted to numpy arr which is then converted to type uint8 in png_converter. I think due to type conversions, we might not be able to assert blobs.
arr = pb2array(d.blob)

The original blob is made from numpy array and the final blob will be made from png after doing URI2Buffer.

Comment (JoanFM) [reply tests/unit/drivers/test_convert_driver.py:28]: Ok have u tested in ur example then that it works at least?

Comment (JoanFM) [reply tests/unit/drivers/test_convert_driver.py:28]: maybe we can have an integration test for this one

Comment (bhavsarpratik) [reply tests/unit/drivers/test_convert_driver.py:28]: Yes tested it for certain values. The test throws error if the uri is empty or there is a huge mismatch of original image shape and final image shape. If original is 32,32 it works for like 32, 64 final but fails for 32, 128. Not able to understand this phenomenon. Throws ValueError: not enough image data.

Let me check integration test. Pushing code for now for testing current.

**Review by JoanFM — COMMENTED** [review]
Where does this env come from?

**Comment by bhavsarpratik** [comment]
I am trying out this 
https://medium.com/ai2-blog/python-caching-in-github-actions-e9452698e98d

File: .github/workflows/ci.yml [hunk 7]
```diff
@@ -35,13 +35,17 @@ jobs:
         uses: actions/setup-python@v1
         with:
           python-version: ${{ matrix.python-version }}
+      - uses: actions/cache@v2
```
Comment (JoanFM) [inline .github/workflows/ci.yml:38]: if we want to have this I have nothing against, but let's put it in a different PR

**Review by JoanFM — APPROVED** [review]
